test/sql/json/json_nested_casts.test

```diff
@@ -95,3 +95,127 @@ query I
 SELECT '42'::UNION(u VARCHAR)::JSON
 ----
 42
+
+# we support our nested to and from JSON (just calls to_json and from_json - well tested)
+query I
+SELECT {duck: 42}::JSON
+----
+{"duck":42}
+
+query I
+SELECT '{"duck":42}'::JSON::STRUCT(duck INTEGER)
+----
+{'duck': 42}
+
+query I
+SELECT ['duck']::JSON
+----
+["duck"]
+
+query I
+SELECT '["duck"]'::JSON::VARCHAR[]
+----
+[duck]
+
+query I
+SELECT MAP(['duck'], [42])::JSON
+----
+{"duck":42}
+
+query I
+SELECT '{"duck":42}'::JSON::MAP(VARCHAR, INTEGER)
+----
+{duck=42}
+
+# casting should be strict
+statement error
+SELECT '{"duck":42}'::JSON::STRUCT(goose INTEGER)
+----
+Conversion Error
+
+statement error
+SELECT '["a", "b", "c"]'::JSON::INT[]
+----
+Conversion Error
+
+statement error
+SELECT '{"duck":42}'::JSON::MAP(INTEGER, INTEGER)
+----
+Conversion Error
+
+# however, when we encounter a conversion error in a nested type with TRY, only that becomes NULL
+query I
+SELECT TRY_CAST('{"duck":42}'::JSON AS STRUCT(goose INTEGER))
+----
+{'goose': NULL}
+
+query I
+SELECT TRY_CAST('["a", "b", "c"]'::JSON AS INT[])
+----
+[NULL, NULL, NULL]
+
+# map keys cannot be NULL
+statement error
+SELECT TRY_CAST('{"duck":42}'::JSON AS MAP(INTEGER, INTEGER))
+----
+Conversion Error
+
+# but values can
+query I
+SELECT TRY_CAST('{"42":"duck"}'::JSON AS MAP(INTEGER, INTEGER))
+----
+{42=NULL}
+
+# we can also cast implicitly to our types
+statement ok
+CREATE TABLE structs (v STRUCT(duck INTEGER))
+
+statement ok
+CREATE TABLE lists (v VARCHAR[])
+
+statement ok
+CREATE table maps (v MAP(VARCHAR, INTEGER))
+
+statement ok
+INSERT INTO structs VALUES ('{"duck":42}'::JSON)
+
+statement ok
+INSERT INTO lists VALUES ('["duck"]'::JSON)
+
+statement ok
+INSERT INTO maps VALUES ('{"duck":42}'::JSON)
+
+query I
+SELECT * FROM structs
+----
+{'duck': 42}
+
+query I
+SELECT * FROM lists
+----
+[duck]
+
+query I
+SELECT * FROM maps
+----
+{duck=42}
+
+# and back to JSON
+statement ok
+CREATE TABLE jsons (j JSON)
+
+statement ok
+INSERT INTO jsons SELECT * FROM structs
+
+statement ok
+INSERT INTO jsons SELECT * FROM lists
+
+statement ok
+INSERT INTO jsons SELECT * FROM maps
+
+query I
+SELECT * FROM jsons
```

Comment:
Could we add more tests with the structs/vectors in `test_all_types` and `test_vector_types` (e.g. `select * from test_vector_types(null::int[], false);`)?

Reply:
I'm happy to add more tests, but I'm not sure I understand what you mean.

`test_all_types` is in DuckDB core, while the `JSON` type is only in the extension. How would this work? Happy to discuss in the office today

Reply:
I just meant including the casts to and from the vectors, e.g.:

```sql
select test_vector::JSON from test_vector_types(null::int[], false);
select test_vector::JSON::int[]::json from test_vector_types(null::int[], false);
select test_vector::JSON from test_vector_types(null::int[][], false);
select test_vector::JSON::int[][]::json from test_vector_types(null::int[][], false);
select test_vector::JSON from test_vector_types(null::row(i varchar, j int), false);
...
```

The `false` indicates whether or not to flatten the vectors, this can be used to find bugs in handling of non-flat vectors by checking if the result is the same for both settings (true and false), e.g.:

```sql
foreach flatten true false

query I nosort expected_result_row
select test_vector::JSON from test_vector_types(null::row(i varchar, j int), ${flatten});

endloop

```

Reply:
This was an excellent suggestion, thanks! I was able to quickly squash a bunch of bugs. I've also added more functionality in the process. We can now cast all of our types to `JSON` and back, e.g.:
```sql
D select '1996-03-27'::DATE::JSON d;
┌──────────────┐
│      d       │
│     json     │
├──────────────┤
│ "1996-03-27" │
└──────────────┘
D select '"1996-03-27"'::JSON::DATE d;
┌────────────┐
│     d      │
│    date    │
├────────────┤
│ 1996-03-27 │
└────────────┘
```

`VARCHAR` is a special case of this, and still behaves like before. It is parsed and validated instead of being interpreted as a string that has to be quoted.